hoomd/hpmc/pytest/test_shape.py

```diff
@@ -0,0 +1,103 @@
+import hoomd
+import hoomd.hpmc
+from hoomd.hpmc import _hpmc
+import numpy as np
+import pytest
+from hoomd.hpmc.pytest.conftest import *
+
+
+def check_dict(shape_dict, args):
+    for key, val in args.items():
+        if isinstance(shape_dict[key], list) and len(shape_dict[key]) > 0 \
+           and key != 'shapes':
+            np.testing.assert_allclose(shape_dict[key], val)
+        elif key == 'shapes':
+            for shape_args in shape_dict[key]:
+                for shape_key, shape_val in shape_args.items():
+                    if isinstance(shape_args[shape_key], list) \
+                       and len(shape_args[shape_key]) > 0:
+                        np.testing.assert_allclose(shape_args[shape_key],
+                                                   shape_val)
+                    else:
+                        assert shape_args[shape_key] == shape_val
+        else:
+            assert shape_dict[key] == val
+
+
+def test_dict_conversion(shape_dict_conversion_args):
+    for shape_params, args_list in shape_dict_conversion_args():
+        for args in args_list:
+            test_shape = shape_params(args)
+            test_dict = test_shape.asDict()
+            check_dict(test_dict, args)
+
+
+def test_shape_params(integrator_args):
+    for shape_integrator, valid_args, invalid_args in integrator_args():
+        mc = shape_integrator(23456)
+        for args in valid_args:
+            mc.shape["A"] = args
+            check_dict(mc.shape["A"], args)
+        for args in invalid_args:
+            with pytest.raises(Exception):
+                mc.shape["A"] = args
+
+
+def test_shape_attached(dummy_simulation_factory, integrator_args):
+    for shape_integrator, valid_args, invalid_args in integrator_args():
+        mc = shape_integrator(23456)
+        for args in valid_args:
+            mc.shape["A"] = args
+            sim = dummy_simulation_factory()
+            sim.operations.add(mc)
+            sim.operations.schedule()
+            check_dict(mc.shape["A"], args)
+
+
+def test_overlaps(device, lattice_simulation_factory, integrator_args):
```

Comment:
I do not see how this effectively tests the shape overlaps. It seems to assume that all valid shape arguments do not overlap in the (0,0,0) / (0,8,0) configuration and do in the (0,0,0) / (0, 0.5,0) configuration.

I'd prefer to see shape spedific hand-constructed tests for this that probe the details of each shape class.

Reply:
Agreed, but that requires a more thorough understanding of the geometry. It may be worth raising an issue about this and dealing with it in a separate pull request. Regardless, it needs to be fixed or documented (for later testing).

Reply:
I just don't see how it is possible to write a generic test for one overlapping configuration and one non-overlapping configuration that is guaranteed to always be correct for all valid shape arguments. Any test that checks for overlaps or not needs to have shape specific knowledge.